handlers/slack/slack.go

```diff
@@ -0,0 +1,390 @@
+package slack
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+	"github.com/pkg/errors"
+)
+
+var apiURL = "https://slack.com/api"
+
+const (
+	configBotToken        = "bot_token"
+	configUserToken       = "user_token"
+	configValidationToken = "verification_token"
+)
+
+var (
+	ErrAlreadyPublic         = "already_public"
+	ErrPublicVideoNotAllowed = "public_video_not_allowed"
+)
+
+func init() {
+	courier.RegisterHandler(newHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+func newHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("SL"), "Slack")}
+}
+
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	s.AddHandlerRoute(h, http.MethodPost, "receive", h.receiveEvent)
+	return nil
+}
+
+func handleURLVerification(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request, payload *moPayload) ([]courier.Event, error) {
+	validationToken := channel.ConfigForKey(configValidationToken, "")
+	if validationToken != payload.Token {
+		w.WriteHeader(http.StatusForbidden)
+		return nil, fmt.Errorf("wrong validation token for channel: %s", channel.UUID())
+	}
+	w.Header().Set("Content-Type", "text/plain")
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte(payload.Challenge))
+	return nil, nil
+}
+
+func (h *handler) receiveEvent(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	payload := &moPayload{}
+	err := handlers.DecodeAndValidateJSON(payload, r)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	if payload.Type == "url_verification" {
+		return handleURLVerification(ctx, channel, w, r, payload)
+	}
+
+	// if event is not a message or is from the bot ignore it
+	if payload.Event.Type == "message" && payload.Event.BotID == "" && payload.Event.ChannelType == "im" {
+
+		date := time.Unix(int64(payload.EventTime), 0)
+
+		userInfo, log, err := getUserInfo(payload.Event.User, channel)
+		if err != nil {
+			h.Backend().WriteChannelLogs(ctx, []*courier.ChannelLog{log})
+			return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+		}
```

Comment:
These changes seem to belong to a method such as `DescribeURN` and that will make sure we are honouring the workspace isAnon status